src/Microsoft.OData.Edm/Csdl/Utf8JsonWriterExtensions.cs

```diff
@@ -0,0 +1,196 @@
+        }
+
+        /// <summary>
+        /// Write the required property (name/value).
+        /// </summary>
+        /// <param name="writer">The JSON writer.</param>
+        /// <param name="name">The property name.</param>
+        /// <param name="value">The property value.</param>
+        public static void WriteRequiredProperty(this Utf8JsonWriter writer, string name, long value)
+        {
+            EdmUtil.CheckArgumentNull(writer, nameof(writer));
+
+            writer.WritePropertyName(name);
+            writer.WriteNumberValue(value);
+        }
+
+        /// <summary>
+        /// Write the required property (name/value).
+        /// </summary>
+        /// <param name="writer">The JSON writer.</param>
+        /// <param name="name">The property name.</param>
+        /// <param name="value">The property value.</param>
+        public static void WriteRequiredProperty(this Utf8JsonWriter writer, string name, bool value)
+        {
+            EdmUtil.CheckArgumentNull(writer, nameof(writer));
+
+            writer.WritePropertyName(name);
+            writer.WriteBooleanValue(value);
+        }
+
+        /// <summary>
+        /// Write the required property (name/value).
+        /// </summary>
+        /// <param name="writer">The JSON writer.</param>
+        /// <param name="name">The property name.</param>
+        /// <param name="value">The property value.</param>
+        public static void WriteRequiredProperty(this Utf8JsonWriter writer, string name, decimal value)
+        {
+            EdmUtil.CheckArgumentNull(writer, nameof(writer));
+
+            writer.WritePropertyName(name);
+            writer.WriteNumberValue(value);
+        }
+
+        /// <summary>
+        /// Write the optional property (name/value).
+        /// </summary>
+        /// <param name="writer">The JSON writer.</param>
+        /// <param name="name">The property name.</param>
+        /// <param name="value">The property value.</param>
+        /// <param name="toString">The value to string Func.</param>
+        public static void WriteRequiredProperty<T>(this Utf8JsonWriter writer, string name, T value, Func<T, string> toString)
+        {
+            EdmUtil.CheckArgumentNull(writer, nameof(writer));
+
+            writer.WritePropertyName(name);
+            writer.WriteStringValue(toString(value));
+        }
+
+        /// <summary>
+        /// Write the optional property (name/value).
+        /// </summary>
+        /// <param name="writer">The JSON writer.</param>
+        /// <param name="name">The property name.</param>
+        /// <param name="value">The property value.</param>
+        public static void WriteOptionalProperty(this Utf8JsonWriter writer, string name, int? value)
+        {
+            EdmUtil.CheckArgumentNull(writer, nameof(writer));
+
+            if (value != null)
+            {
+                writer.WritePropertyName(name);
+                writer.WriteNumberValue(value.Value);

```

Comment:
Should these just call writer.WriteRequiredProperty for non-null values?

Reply:
call "WriteRequiredProperty" again has two disadvantages:
1) check the argument again, (useless)
2) a cost to call extra method. 

Besides, the codes here is clear for what we are doing. It's more readable, thought?

---
In reply to: [442938244](https://github.com/OData/odata.net/pull/1775#discussion_r442938244) [](ancestors = 442938244)

Reply:
Simple enough logic that I'm okay with repeating the code.

---
In reply to: [444553313](https://github.com/OData/odata.net/pull/1775#discussion_r444553313) [](ancestors = 444553313,442938244)